library/core/src/intrinsics.rs

```diff
@@ -968,6 +968,23 @@ pub const unsafe fn assume(b: bool) {
     }
 }
 
+/// Hints to the compiler that current code path is cold.
+///
+/// Note that, unlike most intrinsics, this is safe to call;
+/// it does not require an `unsafe` block.
+/// Therefore, implementations must not require the user to uphold
+/// any safety invariants.
+///
+/// This intrinsic does not have a stable counterpart.
+#[rustc_const_unstable(feature = "const_likely", issue = "none")]
+#[unstable(feature = "core_intrinsics", issue = "none")]
+#[cfg_attr(not(bootstrap), rustc_intrinsic)]
+#[cfg(not(bootstrap))]
+#[rustc_nounwind]
+#[miri::intrinsic_fallback_is_spec]
+#[cold]
```

Comment:
This isn't marked as inline, so can you add a test to check that the call to `cold_path` is properly eliminated from the output IR?

Reply:
I added `CHECK-NOT` to the tests and verified it fails when the call is not eliminated

Reply:
As you have noticed the implementation requires that Rust's own inline pass does not inline an empty cold function, but the LLVM's inline pass does.

At the moment this works. If the behavior changed, we would have to make `cold_path()` a real intrinsic. It would be a no-op that's ignored by all mir passes and backends, but is detected by the `find_cold_blocks()` function.

I didn't implement this change yet in order to keep the pull request small and simple.

Reply:
You can just an an empty implementation of the intrinsic in compiler/rustc_codegen_ssa/src/mir/intrinsic.rs to avoid emitting a call to the `cold_path` function during codegen.

Reply:
I added the empty implementation of `cold_path()` to `compiler/rustc_codegen_ssa/src/mir/intrinsic.rs`

I also removed the `#[rustc_intrinsic]` from `likely` and `unlikely`. They are ordinary functions now. I found that Rust's inliner has several checks to avoid inlining intrinsics. But these functions should really be inlined.